sdk/eventhubs/azure-messaging-eventhubs/src/main/java/com/azure/messaging/eventhubs/EventHubProducerAsyncClient.java

```diff
@@ -502,12 +617,69 @@ private String getEntityPath(String partitionId) {
             : String.format(Locale.US, SENDER_ENTITY_PATH_FORMAT, eventHubName, partitionId);
     }
 
+    private Mono<AmqpSendLink> updatePublishingState(String partitionId, AmqpSendLink amqpSendLink) {
+        if (this.isIdempotentPartitionPublishing) {
+            return amqpSendLink.getRemoteProperties().map(properties -> {
+                this.setPartitionPublishingState(
+                    partitionId, (Long) properties.get(SymbolConstants.PRODUCER_ID),
+                    (Short) properties.get(SymbolConstants.PRODUCER_EPOCH),
+                    (Integer) properties.get(SymbolConstants.PRODUCER_SEQUENCE_NUMBER)
+                );
+                return amqpSendLink;
+            });
+        } else {
+            return Mono.just(amqpSendLink);
+        }
+    }
+
+    /**
+     * Get the idempotent producer's publishing state of a partition from the client side maintained state.
+     * It doesn't create a link to get state from the service.
+     */
+    private PartitionPublishingState getClientPartitionPublishingState(String partitionId) {
+        if (!this.isIdempotentPartitionPublishing) {
+            throw logger.logExceptionAsWarning(
+                new IllegalStateException("getPartitionPublishingState() shouldn't be called if the producer"
+                    + " is not an idempotent producer."));
+        }
+        if (this.partitionPublishingStates.containsKey(partitionId)) {
+            return this.partitionPublishingStates.get(partitionId);
+        } else {
+            synchronized (this) {
```

Comment:
Since we're synchronising on the entire class... is it possible to just synchronise on the parts that need to be?

Reply:
Changed to synchronize on `partitionPublishingStates`.